sample/rules.go

```diff
@@ -190,11 +199,35 @@ func ruleMatchesSpanInTrace(trace *types.Trace, rule *config.RulesBasedSamplerRu
 		return true
 	}
 
+	var seen []int
+
 	for _, span := range trace.GetSpans() {
 		ruleMatched := true
-		for _, condition := range rule.Conditions {
-			// whether this condition is matched by this span.
-			value, exists := extractValueFromSpan(trace, span, condition, checkNestedFields)
+		for i, condition := range rule.Conditions {
+			var value interface{}
+			var exists bool
+			if slices.Contains(seen, i) {
+				continue
+			} else {
+				if strings.Contains(condition.Field, RootPrefix) {
+					value, exists, _ = extractValueFromSpan(trace, span, condition, checkNestedFields)
+					seen = append(seen, i)
+				}
+
+				for _, field := range condition.Fields {
+					if strings.Contains(field, RootPrefix) {
+						value, exists, _ = extractValueFromSpan(trace, span, condition, checkNestedFields)
+						if !slices.Contains(seen, i) { // only append to seen for first field with root prefix
+							seen = append(seen, i)
+						}
+					}
+					// call extractValueFromSpan for each field without root prefix
```

Comment:
I'm sorry, but I don't understand this code. You are calling this function again unconditionally at two different levels, and in none of these cases have the parameters changed at all. This either needs some conditions or some extra assignments. 

This code also doesn't explain `seen` so that I understand its use. Can you please elaborate?

Reply:
Thanks, I think this just need a single else clause to be more concise. And I will add a comment for `seen`.